File: reward_gen.py
"""Turn a natural-language reward description into Python reward code by
shelling out to the Claude Code CLI (`claude -p`), so no API key is needed."""

import json
import os
import re
import subprocess

from envs import ENVS

MODEL = os.environ.get("REWARD_LAB_MODEL", "claude-haiku-4-5-20251001")

PROMPT_TEMPLATE = """You translate a natural-language reward description into a Python reward \
function for a tabular-RL gridworld. Translate LITERALLY: implement exactly what is asked, \
nothing more. Do NOT add penalties, bonuses, shaping, or safeguards the user did not ask for, \
even if you think the description is incomplete or would lead to unintended behavior. \
(This is for an AI-safety demo about misspecified rewards — faithfulness to the spec is the \
whole point.)

The environment is "{env_name}". {tagline}

The function signature is:

    def reward(prev, action, state):

`prev` and `state` are dicts (the state before and after the action). `action` is an int 0-3 \
(up/down/left/right). Available keys in both dicts:

{state_docs}

Counters like the cumulative ones only ever increase; to reward an event once, use deltas, \
e.g. `state['dirt_cleaned'] - prev['dirt_cleaned']`.

The user's reward description:
\"\"\"{spec}\"\"\"

STRICT REQUIREMENT: every reward trigger in the description must come with an explicit number \
("1 point per dirt cleaned", "-5 per vase"). If any trigger has no number ("some points", "a big \
penalty", "reward it for X" with no amount), do NOT write code. Instead reply with ONLY:
{{"error": "<one short sentence saying which trigger is missing a number and showing how to phrase it>"}}

Otherwise reply with ONLY a JSON object (no markdown fences, no prose before or after) of the form:
{{"code": "<the python source of the reward function>", "summary": [{{"icon": "<single emoji>", "text": "<short human-readable line, e.g. '+10 per coin collected'>"}}]}}

The summary must list every term of the reward function, one entry per term, using the exact \
numeric values from the code. Keep the code simple and readable; only use keys listed above."""


class GenError(Exception):
    pass


def build_prompt(env_id, spec):
    env = ENVS[env_id]
    docs = "\n".join(f"  - {k}: {v}" for k, v in env.STATE_DOCS.items())
    return PROMPT_TEMPLATE.format(env_name=env.name, tagline=env.tagline,
                                  state_docs=docs, spec=spec)


def extract_json(text):
    """Pull the first balanced {...} block out of possibly-wrapped output."""
    start = text.find("{")
    if start == -1:
        raise GenError("no JSON in model output")
    depth = 0
    in_str = False
    esc = False
    for i in range(start, len(text)):
        ch = text[i]
        if esc:
            esc = False
            continue
        if ch == "\\":
            esc = True
        elif ch == '"' and not esc:
            in_str = not in_str
        elif not in_str:
            if ch == "{":
                depth += 1
            elif ch == "}":
                depth -= 1
                if depth == 0:
                    return json.loads(text[start:i + 1])
    raise GenError("unbalanced JSON in model output")


def generate_reward(env_id, spec, timeout=90):
    prompt = build_prompt(env_id, spec)
    try:
        out = subprocess.run(
            ["claude", "-p", prompt, "--model", MODEL, "--output-format", "json"],
            capture_output=True, text=True, timeout=timeout, cwd="/tmp",
        )
    except subprocess.TimeoutExpired:
        raise GenError("Claude took too long to respond")
    except FileNotFoundError:
        raise GenError("`claude` CLI not found on PATH")
    if out.returncode != 0:
        raise GenError(f"claude CLI failed: {out.stderr[:300]}")
    try:
        wrapper = json.loads(out.stdout)
        result_text = wrapper.get("result", "")
    except json.JSONDecodeError:
        result_text = out.stdout
    data = extract_json(result_text)
    if data.get("error"):
        raise GenError(data["error"])
    code = data.get("code", "")
    summary = data.get("summary", [])
    if "def reward" not in code:
        raise GenError("model did not return a reward function")
    # normalize escaped newlines if the model double-escaped
    if "\\n" in code and "\n" not in code:
        code = code.replace("\\n", "\n")
    return {"code": code, "summary": summary}

File: rewardgen.js
/* Turn a natural-language reward description into a JavaScript reward function
 * by calling OpenRouter directly from the browser with the user's own API key.
 * The key never leaves the browser except to go to openrouter.ai. Ported from
 * reward_gen.py. */
(function (root) {
  const ENVS = root.REWARDLAB_ENVS;
  const DEFAULT_MODEL = "anthropic/claude-haiku-4.5";

  const PROMPT_TEMPLATE = `You translate a natural-language reward description into a JavaScript reward \
function for a tabular-RL gridworld. Translate LITERALLY: implement exactly what is asked, \
nothing more. Do NOT add penalties, bonuses, shaping, or safeguards the user did not ask for, \
even if you think the description is incomplete or would lead to unintended behavior. \
(This is for an AI-safety demo about misspecified rewards — faithfulness to the spec is the \
whole point.)

The environment is "{env_name}". {tagline}

The function signature is:

    function reward(prev, action, state) { ... }

\`prev\` and \`state\` are plain objects (the state before and after the action). \`action\` is an \
integer 0-3 (up/down/left/right). The function must return a number. \`Math\` is available. \
Available keys in both \`prev\` and \`state\`:

{state_docs}

Counters like the cumulative ones only ever increase; to reward an event once, use deltas, \
e.g. \`state.dirt_cleaned - prev.dirt_cleaned\`. Booleans are true/false.

STRICT REQUIREMENT: every reward trigger in the description must come with an explicit number \
("1 point per dirt cleaned", "-5 per plant"). If any trigger has no number ("some points", "a big \
penalty", "reward it for X" with no amount), do NOT write code. Instead reply with ONLY:
{"error": "<one short sentence saying which trigger is missing a number and showing how to phrase it>"}

Otherwise reply with ONLY a JSON object (no markdown fences, no prose before or after) of the form:
{"code": "<the javascript source of the reward function, starting with 'function reward'>", "summary": [{"icon": "<single emoji>", "text": "<short human-readable line, e.g. '+10 per coin collected'>"}]}

The summary must list every term of the reward function, one entry per term, using the exact \
numeric values from the code. Keep the code simple and readable; only use keys listed above.`;

  function buildPrompt(envId, spec) {
    const meta = ENVS.envMeta().find((e) => e.id === envId);
    const docs = Object.entries(meta.state_docs).map(([k, v]) => `  - ${k}: ${v}`).join("\n");
    return PROMPT_TEMPLATE
      .replace("{env_name}", meta.name)
      .replace("{tagline}", meta.tagline)
      .replace("{state_docs}", docs);
  }

  // Pull the first balanced {...} block out of possibly-wrapped output.
  function extractJSON(text) {
    const start = text.indexOf("{");
    if (start === -1) throw new Error("no JSON in model output");
    let depth = 0, inStr = false, esc = false;
    for (let i = start; i < text.length; i++) {
      const ch = text[i];
      if (esc) { esc = false; continue; }
      if (ch === "\\") esc = true;
      else if (ch === '"') inStr = !inStr;
      else if (!inStr) {
        if (ch === "{") depth++;
        else if (ch === "}") { depth--; if (depth === 0) return JSON.parse(text.slice(start, i + 1)); }
      }
    }
    throw new Error("unbalanced JSON in model output");
  }

  async function generateReward(envId, spec, apiKey, model) {
    if (!apiKey) throw new Error("enter your OpenRouter API key first");
    const prompt = buildPrompt(envId, spec) + "\n\nThe user's reward description:\n\"\"\"" + spec + "\"\"\"";
    let resp;
    try {
      resp = await fetch("https://openrouter.ai/api/v1/chat/completions", {
        method: "POST",
        headers: {
          "Authorization": "Bearer " + apiKey,
          "Content-Type": "application/json",
          "HTTP-Referer": root.location ? root.location.origin : "https://reward-lab",
          "X-Title": "Reward Lab",
        },
        body: JSON.stringify({
          model: model || DEFAULT_MODEL,
          max_tokens: 1024,
          messages: [{ role: "user", content: prompt }],
        }),
      });
    } catch (e) {
      throw new Error("OpenRouter request failed: " + e.message);
    }
    if (!resp.ok) {
      const body = await resp.text().catch(() => "");
      throw new Error("OpenRouter error " + resp.status + ": " + body.slice(0, 300));
    }
    const out = await resp.json();
    let content;
    try { content = out.choices[0].message.content; }
    catch (e) { throw new Error("unexpected OpenRouter response: " + JSON.stringify(out).slice(0, 300)); }

    const data = extractJSON(content);
    if (data.error) throw new Error(data.error);
    let code = data.code || "";
    const summary = data.summary || [];
    if (!/function\s+reward/.test(code) && !/reward\s*=/.test(code))
      throw new Error("model did not return a reward function");
    if (code.includes("\\n") && !code.includes("\n")) code = code.replace(/\\n/g, "\n");
    return { code, summary };
  }

  root.REWARDLAB_GEN = { generateReward, DEFAULT_MODEL };
})(typeof window !== "undefined" ? window : globalThis);
